File: backend/search_agent/extraction/social.py
"""Social profile link extraction (deterministic, no LLM)."""

import re

SOCIAL_PATTERNS: list[tuple[str, re.Pattern]] = [
    ("linkedin", re.compile(r"https?://(?:[\w\-]+\.)?linkedin\.com/[^\s\"'<>)]+", re.I)),
    ("facebook", re.compile(r"https?://(?:[\w\-]+\.)?facebook\.com/[^\s\"'<>)]+", re.I)),
    ("instagram", re.compile(r"https?://(?:[\w\-]+\.)?instagram\.com/[^\s\"'<>)]+", re.I)),
    ("x", re.compile(r"https?://(?:[\w\-]+\.)?(?:twitter|x)\.com/[^\s\"'<>)]+", re.I)),
]


def extract_social_links(html_or_text: str) -> list[dict]:
    """Return [{"platform": ..., "url": ...}] for known social networks."""
    links: list[dict] = []
    seen: set[str] = set()
    for platform, pattern in SOCIAL_PATTERNS:
        for url in pattern.findall(html_or_text):
            url = url.rstrip(".,;")
            if url not in seen:
                seen.add(url)
                links.append({"platform": platform, "url": url})
    return links

File: src/search-agent/extraction/social.ts
export type SocialPlatform = "linkedin" | "facebook" | "instagram" | "x" | "other";

const SOCIAL_PATTERNS: Array<{ platform: SocialPlatform; regex: RegExp }> = [
  { platform: "linkedin", regex: /https?:\/\/(?:[\w.-]+\.)?linkedin\.com\/[^\s"'<>]+/gi },
  { platform: "facebook", regex: /https?:\/\/(?:[\w.-]+\.)?facebook\.com\/[^\s"'<>]+/gi },
  { platform: "instagram", regex: /https?:\/\/(?:[\w.-]+\.)?instagram\.com\/[^\s"'<>]+/gi },
  { platform: "x", regex: /https?:\/\/(?:[\w.-]+\.)?(?:twitter|x)\.com\/[^\s"'<>]+/gi },
];

export function extractSocialLinks(
  htmlOrText: string,
): Array<{ platform: SocialPlatform; url: string }> {
  const links: Array<{ platform: SocialPlatform; url: string }> = [];
  const seen = new Set<string>();

  for (const { platform, regex } of SOCIAL_PATTERNS) {
    const matches = htmlOrText.match(regex) ?? [];
    for (const url of matches) {
      const cleaned = url.replace(/[.,;)]+$/, "");
      if (!seen.has(cleaned)) {
        seen.add(cleaned);
        links.push({ platform, url: cleaned });
      }
    }
  }

  return links;
}
